stbt/android.py

```diff
@@ -0,0 +1,407 @@
+
+        x, y = self._to_native_coordinates(x, y)
+        self.adb(["shell", "input", "tap", str(x), str(y)], timeout_secs=10)
+
+    def _adb(self, command, timeout_secs=None, **kwargs):
+        _command = []
+        if timeout_secs is not None:
+            _command += ["timeout", "%fs" % timeout_secs]
+        _command += [self.adb_binary]
+        if self.adb_server:
+            _command += ["-H", self.adb_server]
+        if self.adb_device:
+            _command += ["-s", self.adb_device]
+        _command += command
+        stbt.debug("AdbDevice.adb: About to run command: %r\n" % _command)
+        output = subprocess.check_output(
+            _command, stderr=subprocess.STDOUT, **kwargs)
+        if command[0] in ("connect", "tcpip", "usb"):
+            time.sleep(5)
+        return output
+
+    def _connect(self, timeout_secs):
+        if not self.adb_device:
+            raise RuntimeError('AdbDevice: error: If "tcpip=True" '
+                               'you must specify "adb_device"')
+        try:
+            if self.adb_device in self._adb(["devices"]):
+                return
+        except subprocess.CalledProcessError:
+            pass
+
+        # "adb connect" always returns success; we have to parse the output
+        # which looks like "connected to 192.168.2.163:5555" or
+        # "already connected to 192.168.2.163:5555" or
+        # "unable to connect to 192.168.2.100:5555".
+        output = self._adb(["connect", self.adb_device], timeout_secs)
+        if ("connected to %s" % self.adb_device) not in output:
+            sys.stderr.write(output)
+            raise AdbError(0, "adb connect %s" % self.adb_device, output, self)
+
+    def _to_native_coordinates(self, x, y):
+        if self.coordinate_system == CoordinateSystem.ADB_NATIVE:
+            return x, y
+        else:
+            return _to_native_coordinates(
+                x, y, self.coordinate_system, self._get_display_dimensions())
+
+    def _get_display_dimensions(self):
+        return _parse_display_dimensions(
+            self.adb(["shell", "dumpsys", "window"],
+                     timeout_secs=10, capture_output=True))
+
+
+class AdbError(Exception):
+    def __init__(self, returncode, cmd, output=None, adb_control=None):
+        super(AdbError, self).__init__()
+        self.returncode = returncode
+        self.cmd = cmd
+        self.output = output
+        self.adb_devices = None
+        if adb_control:
+            self.adb_devices = adb_control.devices()
+
+    def __str__(self):
+        return "Command '%s' failed with exit status %d. Output:\n%s\n%s" % (
+            self.cmd, self.returncode, self.output, self.adb_devices)
+
+
+def _region_to_tuple(r):
+    if isinstance(r, stbt.Region):
+        return ((r.x + r.right) // 2, (r.y + r.bottom) // 2)
+    elif isinstance(r, tuple) and len(r) == 2:
+        return r
+    else:
+        raise ValueError("Expected stbt.Region or (x,y) tuple but got %r" % r)
```

Comment:
I think this should be `TypeError`.

An alternative implementation that takes advantage of Python's duck typing might look like:

```
try:
    return (r.x + getattr(r, "width", 0) // 2, r.y + getattr(r, "height", 0) // 2)
except AttributeError:
    pass
try:
    x, y = r
    return (x, y)
except ValueError:
    raise TypeError("Expected stbt.Region or (x, y) coordinates but got %r", r)
```